external/xtorch_harmonics/xtorch_harmonics/xtorch_harmonics.py

```diff
@@ -0,0 +1,208 @@
+    (LOBATTO_GRID, LOBATTO_GRID),
+    (EQUIANGULAR_GRID, LEGENDRE_GAUSS_GRID),
+    (LEGENDRE_GAUSS_GRID, EQUIANGULAR_GRID),
+}
+QUADRATURE_FUNCTIONS = {
+    EQUIANGULAR_GRID: torch_harmonics.quadrature.clenshaw_curtiss_weights,
+    LEGENDRE_GAUSS_GRID: torch_harmonics.quadrature.legendre_gauss_weights,
+    LOBATTO_GRID: torch_harmonics.quadrature.lobatto_weights,
+}
+LON_DIM = "grid_xt"
+LAT_DIM = "grid_yt"
+
+T_XarrayObject = TypeVar("T_XarrayObject", xr.DataArray, xr.Dataset)
+T_Grid = Literal[EQUIANGULAR_GRID, LEGENDRE_GAUSS_GRID, LOBATTO_GRID]
+
+
+def compute_quadrature_latitudes(n_lat: int, grid: T_Grid) -> np.ndarray:
+    if grid not in QUADRATURE_FUNCTIONS:
+        raise ValueError(f"Unrecognized grid {grid!r}.")
+
+    quadrature_function = QUADRATURE_FUNCTIONS[grid]
+    nodes, _ = quadrature_function(n_lat)
+    return np.rad2deg(np.arcsin(nodes))
+
+
+def compute_quadrature_longitudes(n_lon: int) -> np.ndarray:
+    bounds = np.linspace(0, 360, n_lon + 1, endpoint=True)
+    return (bounds[:-1] + bounds[1:]) / 2
+
+
+def _validate_quadrature_points(
+    obj: T_XarrayObject, forward_grid: T_Grid, lat_dim: Hashable, lon_dim: Hashable
+):
+    n_lat = obj.sizes[lat_dim]
+    n_lon = obj.sizes[lon_dim]
+
+    expected_latitudes = compute_quadrature_latitudes(n_lat, forward_grid)
+    expected_longitudes = compute_quadrature_longitudes(n_lon)
+
+    try:
+        if lat_dim in obj.coords:
+            np.testing.assert_allclose(obj[lat_dim].data, expected_latitudes)
+            # TODO: could warn or raise if lat_dim not in obj.coords.
+    except AssertionError:
+        raise ValueError(
+            f"Latitude coordinate {obj[lat_dim]} does not match the expected "
+            f"quadrature points for the provided forward grid {forward_grid!r}."
+        )
+
+    try:
+        if lon_dim in obj.coords:
+            # TODO: I suppose for this check we really only need to ensure that
+            # points are equally spaced and global.  Whether they exactly align
+            # with a regular grid from 0 to 360 degrees E is not necessarily
+            # relevant.
+            np.testing.assert_allclose(obj[lon_dim].data, expected_longitudes)
+            # TODO: could warn or raise if lon_dim not in obj.coords.
+    except AssertionError:
+        raise ValueError(
+            f"Longitude coordinate {obj[lon_dim]} does not match the expected "
+            f"quadrature points for the provided forward grid {forward_grid!r}."
+        )
+
+
+def _roundtrip_numpy(
+    array: np.array, forward_grid: T_Grid, inverse_grid: T_Grid
+) -> np.ndarray:
+    *_, n_lat, n_lon = array.shape
+    tensor = torch.tensor(array).type(torch.double)
+    forward_transform = torch_harmonics.RealSHT(n_lat, n_lon, grid=forward_grid)
+    inverse_transform = torch_harmonics.InverseRealSHT(n_lat, n_lon, grid=inverse_grid)
+    roundtripped = inverse_transform(forward_transform(tensor))
+    return np.array(roundtripped.type(torch.float))
```

Comment:
Although torch harmonics requires input to be double precision, not sure we always want to convert back to float. Can you do something like `roundtripped.astype(array.type)` where array is the input numpy array?

Reply:
Good catch!  Yes, that is important.  I modified things accordingly.  I think it's easier to handle once it is back as a NumPy array.